### ISSUE TITLE
Try gradient accumulation and checkpointing for Linux training on smaller GPUs

### ISSUE THREAD

**Issue by bbrowning** [instructlab/instructlab#2103]
https://huggingface.co/docs/transformers/v4.38.2/perf_train_gpu_one#gradient-accumulation

In the `TrainingArguments` passed to `SFTTrainer`, we can likely reduce the total GPU memory required to train on Linux by enabling gradient accumulation and checkpointing. This can sometimes meaningfully reduce total GPU memory required, with a tradeoff in overall performance.

I don't yet have a GPU to test this on myself, but perhaps some of the other GPU training on Linux testers can poke at this. The code already checks the free GPU memory and warns the user if they don't have enough - we could perhaps turn that from a pure warning into enabling of some of these options automatically if the memory requires are substantially lessened with them enabled.

**Comment by jaideepr97** [comment]
Since this is clearly focused on the laptop use case - do we have the ability to achieve this via what is present in `instructlab/training` already?
If not, I might suggest moving this issue to that repo so that the library is updated to handle this use case and we can probably stop supporting the legacy training route?

If we do decide to continue supporting it - as a separate issue I might suggest moving legacy training into the training repo anyway so all the various training workflows can live in a single place, and this issue can be moved there as is


how would people feel about that?

**Comment by RobotSail** [comment]
@jaideepr97 This issue is related to a separate codepath altogether. The legacy training mechanism is not a part of the training library (yet) but rather a part of the CLI repository. We have all decided that we do not intend to support new updates for the legacy training, and are only keeping it in ilab for the purpose of allowing CPU-only training for a number of use-cases. 

For this to be applicable to the training library, we'd first need to move legacy training there.

**Comment by nathan-weinberg** [comment]
Worth noting there is an issue to track that migration already: https://github.com/instructlab/training/issues/117

In my humble opinion we should prioritize that - the longer training is split between the two, the more problems like this I think will arise.

**Comment by jaideepr97** [comment]
> @jaideepr97 This issue is related to a separate codepath altogether. The legacy training mechanism is not a part of the training library (yet) but rather a part of the CLI repository. We have all decided that we do not intend to support new updates for the legacy training, and are only keeping it in ilab for the purpose of allowing CPU-only training for a number of use-cases.
> 
> For this to be applicable to the training library, we'd first need to move legacy training there.

I'm aware they are separate code paths. What I'm asking is, does the library support CPU-only training today? If not maybe we should be adding that capability there first. Then this issue could be an optimization to that implementation in the library and we won't have any need to continue supporting legacy training anywhere (either CLI or the library) since the benefit it provides will then be covered by the library

**Comment by RobotSail** [comment]
@jaideepr97 The training library only exists training full-size safetensors models today, whereas legacy Linux training only supports GGUF training with llama-cpp, so they are pretty mutually exclusive in terms of what they can train. We'll need to move over a lot of things into the training library to make this work.

I agree with @nathan-weinberg that this should be done sooner than later.

What I'm saying is:
- This issue currently applies only to the https://github.com/instructlab/instructlab repository as this is where legacy linux training (GGUF model training with lora) currently lives
- BUT we aren't actively supporting this training format, hence the name
- However, we want the training library to provide CPU training and other formats, so we would need this code to move over there
- Once the legacy training code has moved over to the training library (https://github.com/instructlab/training), this issue should be moved over there
